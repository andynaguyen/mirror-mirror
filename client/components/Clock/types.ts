interface ClockProps {
  data?: ClockData;
  visible: boolean;
}

interface ClockData {
  date: string;
  day: string;
  time: string;
}

export { ClockProps, ClockData };
