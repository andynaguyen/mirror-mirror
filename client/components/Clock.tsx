import moment from 'moment';
import * as React from 'react';
import styled from 'styled-components';
import withRefresh from '../refresher';

const DATE_FORMAT = 'MMM Do, YYYY';
const DAY_FORMAT = 'dddd';
const TIME_FORMAT = 'hh:mm:ss';
const REFRESH_INTERVAL = 1000;

const Date = styled.div`
  display: flex;
  font-size: 18px;
  justify-content: flex-end;
`;

const Day = styled.div`
  display: flex;
  font-size: 18px;
  justify-content: flex-end;
`;

const Time = styled.div`
  display: flex;
  font-size: 48px;
  font-family: 'Raleway';
  justify-content: flex-end;
  margin-bottom: 12px;
`;

interface ClockProps {
  date: string;
  day: string;
  time: string;
  visible: boolean;
}

const Clock: React.SFC<ClockProps> = ({ date, day, time, visible }) =>
  visible ? (
    <>
      <Time>{time}</Time>
      <Day>{day}</Day>
      <Date>{date}</Date>
    </>
  ) : null;

/**
 * Packages the date, day and time of the moment variable into an object.
 *
 * @param m moment object to format
 */
const formatMoment = (m: moment.Moment) => ({
  date: m.format(DATE_FORMAT),
  day: m.format(DAY_FORMAT),
  time: m.format(TIME_FORMAT),
});

const options = {
  defaultState: formatMoment(moment(0)),
  interval: REFRESH_INTERVAL,
  refresh: () => Promise.resolve(formatMoment(moment())),
};

/**
 * A clock component that updates the date, day, and time every second.
 */
const EnhancedClock = withRefresh(options, Clock);
export default EnhancedClock;
