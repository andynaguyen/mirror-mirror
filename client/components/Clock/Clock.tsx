import * as React from 'react';
import styled from 'styled-components';
import { ClockProps } from './types';

const Date = styled.div`
  display: flex;
  font-size: 18px;
  justify-content: flex-end;
`;

const Day = styled.div`
  display: flex;
  font-size: 18px;
  justify-content: flex-end;
`;

const Time = styled.div`
  display: flex;
  font-size: 48px;
  font-family: 'Raleway';
  justify-content: flex-end;
  margin-bottom: 12px;
`;

const Clock: React.SFC<ClockProps> = ({ data, visible }) =>
  visible ? (
    <>
      <Time>{data.time}</Time>
      <Day>{data.day}</Day>
      <Date>{data.date}</Date>
    </>
  ) : null;

export default Clock;
