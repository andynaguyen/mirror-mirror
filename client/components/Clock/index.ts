import moment from 'moment';
import * as React from 'react';
import Clock from './Clock';
import withRefresh, { RefreshOptions } from '../../refresher';
import { ClockData } from './types';

const DATE_FORMAT = 'MMM Do, YYYY';
const DAY_FORMAT = 'dddd';
const TIME_FORMAT = 'hh:mm:ss';
const REFRESH_INTERVAL = 1000;

/**
 * Packages the date, day and time of the moment variable into an object.
 *
 * @param m moment object to format
 */
const formatMoment = (m: moment.Moment): ClockData => ({
  date: m.format(DATE_FORMAT),
  day: m.format(DAY_FORMAT),
  time: m.format(TIME_FORMAT),
});

const options: RefreshOptions<ClockData> = {
  defaultState: formatMoment(moment(0)),
  interval: REFRESH_INTERVAL,
  refresh: () => Promise.resolve(formatMoment(moment())),
};

/**
 * A clock component that updates the date, day, and time every second.
 */
const EnhancedClock = withRefresh(options, Clock);
export default EnhancedClock;
