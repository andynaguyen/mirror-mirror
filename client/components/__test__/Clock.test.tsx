import 'jest-styled-components';
import * as React from 'react';
import * as renderer from 'react-test-renderer';
import Clock from '../Clock';

describe('<Clock />', () => {
  it('should render the contents of the clock component', () => {
    const component = renderer.create(<Clock visible />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render nothing', () => {
    const component = renderer.create(<Clock visible={false} />);
    const tree = component.toJSON();
    expect(tree).toBeNull();
  });
});
