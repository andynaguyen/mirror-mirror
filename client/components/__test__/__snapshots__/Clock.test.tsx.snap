// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Clock /> should render the contents of the clock component 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 18px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 18px;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 48px;
  font-family: 'Raleway';
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin-bottom: 12px;
}

<div>
  <div
    className="c0"
  >
    04:00:00
  </div>
  <div
    className="c1"
  >
    Wednesday
  </div>
  <div
    className="c2"
  >
    Dec 31st, 1969
  </div>
</div>
`;
